src/units/abilities.cpp

```diff
@@ -1168,52 +1168,136 @@ void attack_type::modified_attacks(unsigned & min_attacks,
+	for(auto& i : damage_type_list) {
+		const config& c = *i.ability_cfg;
+		if(!c["alternative_type"].empty()) {
+			std::string type = c["alternative_type"].str();
+			if(!type_res[type]){
+				type_res[type] = u.resistance_value(resistance_abilities, type);
+				max_res = max_res ? std::max(*max_res, *type_res[type]) : *type_res[type];
+			}
+		}
+	}
+
+	if (type_res.empty()) return "";
+
+	std::vector<std::string> type_list;
+	if(max_res){
+		for(auto& i : type_res){
+			if(i.second == *max_res){
+				type_list.push_back(i.first);
 			}
-			std::sort( std::begin( type_list ) , std::end( type_list ) , [&]( const std::string& rhs , const std::string& lhs ){
-				return type_count[lhs] < type_count[rhs];
-			});
 		}
+		std::sort(type_list.begin(), type_list.end());
 	}
+
 	if(!type_list.empty()){
 		return type_list.front();
 	}
-	return std::nullopt;
+	return "";
+}
+
+std::string attack_type::select_damage_type(const unit_ability_list& damage_type_list, const std::string& key_name, unit_ability_list resistance_abilities) const
+{
+	bool is_alternative = (key_name == "alternative_type");
+	if(is_alternative && other_){
+		return select_alternative_type(damage_type_list, resistance_abilities, (*other_));
+	} else if(!is_alternative){
+		return select_replacement_type(damage_type_list);
+	}
+	return "";
 }
 
 /**
  * Returns the type of damage inflicted.
  */
 std::pair<std::string, std::string> attack_type::damage_type() const
 {
-	unit_ability_list abil_list = get_specials_and_abilities("damage_type");
-	if(abil_list.empty()){
+	unit_ability_list damage_type_list = get_specials_and_abilities("damage_type");
+	if(damage_type_list.empty()){
 		return {type(), ""};
 	}
 
-	std::optional<std::string> replacement_type = select_damage_type(abil_list, "replacement_type");
-	std::optional<std::string> alternative_type = select_damage_type(abil_list, "alternative_type");
-	std::string type_damage = replacement_type.value_or(type());
-	if(alternative_type && type_damage != *alternative_type){
-		return {type_damage, *alternative_type};
+	unit_ability_list resistance_abilities;
+	if(other_){
+		resistance_abilities = (*other_).get_abilities_weapons("resistance", other_loc_, other_attack_, shared_from_this());
+	}
+	std::string replacement_type = select_damage_type(damage_type_list, "replacement_type", resistance_abilities);
+	std::string alternative_type = select_damage_type(damage_type_list, "alternative_type", resistance_abilities);
+	std::string type_damage = replacement_type.empty() ? type() : replacement_type;
+	if(!alternative_type.empty() && type_damage != alternative_type){
+		return {type_damage, alternative_type};
 	}
 	return {type_damage, ""};
 }
 
+std::vector<std::string> attack_type::damage_alternative_type() const
+{
+	unit_ability_list damage_alternative_type_list = get_specials_and_abilities("damage_type");
+	if(damage_alternative_type_list.empty()){
+		return {};
+	}
+	std::unordered_map<std::string, unsigned int> type_count;
```

Comment:
This should be a set. The return type presumably as well.

Reply:
changed map to set ut let vector for sort that by alphabetical order